Add support for Mythic Beasts API v2

lexitus opened the pull request:
Hi,

I've attempted to add support for Mythic Beasts' v2 DNS API. Currently it is passing the test suite, and I am using it with dehydrated to create certs, so it is somewhat functional.

I am not an experienced python developer, so apologies in advance - things I am unsure of : 

- Mythic Beasts does not provide an id for records - I am creating one using a hash.
- There is a loop in _create_record where I check to see if the record has gone live yet. This causes a lot of spam in the recording, which I don't know if it is a problem or not.
- Probably a lot of non-idiomatic things.

adferrand commented on tests/fixtures/cassettes/mythicbeasts/IntegrationTests/test_provider_authenticate.yaml:
```diff
@@ -0,0 +1,80 @@
+    uri: https://auth.mythic-beasts.com/login
+  response:
+    body:
+      string:
```
Hello @lexitus, and thanks a lot for your PR!

The tests are failing because the content of the `response.body.string` parameters in the VCRpy cassettes are objects, when it is supposed to be strings. For the record, VCRpy register any HTTP exchanges during the tests in files, so they can be replayed later without the need to connect to the actual DNS provider API.

How have you registered these cassettes?

lexitus replied:
Thanks for taking a look. :-)

I am trying to follow the docs for recording the cassettes. It seems I was processing the response incorrectly in _filter_response where I was removing the access token, so the response got saved wrong. The output looks more reasonable now.

adferrand replied:
Indeed, now the tests are fixed and are passing.

adferrand commented on lexicon/providers/mythicbeasts.py:
```diff
@@ -0,0 +1,280 @@
+    def _authenticate(self):
+        payload = self._get("/zones")
+
+        if self.domain is None:
```
In fact by construction, `self.domain` is always set by construction in the parent class (see https://github.com/AnalogJ/lexicon/blob/59419ffba7797a860251c83f73221054b3f72862/lexicon/providers/base.py#L62).

The intention in `_authenticate()` is doing the following things:
1) retrieve the credentials required to authenticate from the provider options
2) authenticate against the API and retrieve the tokens that will be used for the lifetime of the provider instance for all HTTP calls
3) verify that with these tokens we are authorized to manipulate the DNS zone targeted by `self.domain`
4) set the internal domain id on `self.domain_id` for the lifetime of the provider instance

So basically I think that here you should make the autentication token retrieval (if not set by the caller), then ensure that you can see the target zone specified by `self.domain`, then assign its value to `self.domain_id` (you do not have necessarily to hash it, it can be the domain itself).

adferrand commented on lexicon/providers/mythicbeasts.py:
```diff
@@ -0,0 +1,280 @@
+        try:
+            payload = self._post(f"/zones/{self.domain}/records", data)
+        except requests.exceptions.HTTPError as err:
+            if err.response.status_code != 400 and err.response.json()["errors"][0][0:16] != 'Duplicate record':
```
Maybe you could add a log debug entry here to say that the record already exists.

adferrand commented on lexicon/providers/mythicbeasts.py:
```diff
@@ -0,0 +1,280 @@
+            timeout = 300
+            start = time.time()
+
+            while(not changes_are_live and time.time() - start < timeout):
```
Your retry loop is totally fine. But if you want to leverage some of the complexity to a built-in module, you will find examples of how to handle a retrying module with `urllib3` in dnsmadeasy and godaddy providers (in the `_request` method).

One interesting thing with urllib3 `Retry` is that you can have a loop where the wait between two attempts start with a low value that increase progressively, increasing changes that the create record will be validated faster that with a fixed wait time.

adferrand commented on lexicon/providers/mythicbeasts.py:
```diff
@@ -0,0 +1,280 @@
+                "ttl": record["ttl"],
+                "content": record["data"],
+                # no id is available, so we need to make our own
+                "id": hashlib.md5(
```
Here also I can propose you a code to generate a "short" indentifier that is used in several modules where ids are not provided by the API: https://github.com/AnalogJ/lexicon/blob/59419ffba7797a860251c83f73221054b3f72862/lexicon/providers/azure.py#L361-L365

adferrand reviewed (changes requested):
Ok this is a great implementation! I have some propositions to align the provider behavior with the other providers in some specific points. Thanks!

lexitus commented:
Thanks for the detailed feedback. I've tried to follow your suggestions. 

For some reason one of the tests failed - I don't really understand why. It looks like it happens during some of the setup before it gets to my code?

adferrand reviewed (changes requested):
After merging master, the tests are passing on all environments. I also took the occasion to remove most of the `409` responses in the cassettes, so that the simulated API returns `200` immediately and the tests are passing immediately instead of 40 minutes. I kept one `409` response to continue to test the retry mechanism.

As a final modification before merging, could you add Mythic Beasts in the supported providers in the README.rst file, and remove it from the potential providers in the developer guide?

lexitus commented:
Thanks - good idea to prune the 409s out of the cassettes 👍

I've updated the docs as requested, and also added myself to CODEOWNERS. One of the tests failed again for no obvious reason...

adferrand reviewed (approved):
I relaunched the CI and builds are passing. So LGTM!
